### ISSUE TITLE
Starting a docker image app fails if it tries to establish connections before Envoy is ready

### ISSUE THREAD

**Issue by stoyanr** [cloudfoundry/cf-for-k8s#189]
### Summary

Starting or restarting a docker image app fails if the app tries to establish connections before the Envoy proxy is ready, and this communication failure causes the app to crash.

```
$ cf restart service-manager
Restarting app service-manager in org SAP_servicemanager / space service-manager as admin...

Stopping app...

Waiting for app to start...
Start unsuccessful
```

Examining the corresponding pod shows that the `opi` container was restarted once due to an error. The logs from the first (failed) run are as follows:

```
{"component_type":"application","correlation_id":"-","level":"info","logger":"sm/sm.go:105","msg":"Setting up Service Manager storage...","type":"log","written_at":"2020-05-08T09:23:04.068924213Z","written_ts":"1588929784"}
panic: unable to construct service-manager builder: error opening storage: error opening storage: could not update database schema: dial tcp 100.65.158.99:5432: connect: connection refused

goroutine 1 [running]:
main.main()
	/go/src/github.wdf.sap.corp/SvcManager/sm-sap/main.go:34 +0x783
```

The above logs indicate that the app tried to connect to its database, but the connection was refused, which caused the app to crash. The reason that the connection was refused is that by that time (2020-05-08T09:23:04), the Envoy proxy hasn't started yet, since the `istio-proxy` logs indicate that this happens a second later:

```
2020-05-08T09:23:05.767841Z	info	Envoy proxy is ready
```

After the `opi` container is restarted by Kubernetes, the app starts successfully and connects to its database. So the end result is a running app, but a failed `cf start`.

This is reproducible with several apps on our side. All of them try to establish connections before Envoy has started. Some of them treat the "connection refused" error as fatal and crash, leading to the above error. Others treat it as non-fatal or retry, and so avoid the error.

### Reproduction Steps

Push an app from a docker image that is fairly quick to start, tries to establish a connection immediately after starting, and treats the failure as fatal. For best effect, push it with `--no-start` and then start it with `cf start`, or restart it with `cf restart`.

### Expected behavior

`cf start` or `cf restart` don't return an error in this case. This could be achieved by either ensuring that Envoy proxy starts before the app tries to establish a connection, or treating the first crash from the app as non-fatal.

**Comment by loewenstein** [comment]
Currently, cf-k8s-networking shouldn't do anything on the egress path from an application to anywhere else, should it?

If that is true, we could use a `Sidecar` resource to configure Istio to not capture the outgoing traffic at all (or does the `Sidecar` resource apply only after the iptable rules are already setup to route any traffic through the Envoy?

**Comment by tcdowney** [comment]
We (CF App Connectivity team) have a story to investigate this behavior for https://github.com/cloudfoundry/cf-for-k8s/issues/167. I added this Issue to it as well: [Tracker Story #172627350](https://www.pivotaltracker.com/story/show/172627350)

We should have a better understanding after we've investigated it on that story and will get back to you.

> or does the Sidecar resource apply only after the iptable rules are already setup to route any traffic through the Envoy

https://preliminary.istio.io/blog/2019/data-plane-setup/

The `istio-init` init container sets up the iptable rules before the app container / Envoy sidecar container are started. The `Sidecar` resource just configures that Envoy. I believe what we're seeing here happens because the app container is running before the Envoy sidecar is ready. 

We really need first class sidecar container support / startup ordering...
https://github.com/kubernetes/enhancements/issues/753

This is all conjecture, though. We'll report back after [ #172627350](https://www.pivotaltracker.com/story/show/172627350).

**Comment by lucaschimweg** [comment]
We have created a minimal [reproduction setup](https://gist.github.com/lucaschimweg/2d1d93aa81b5a4f998ecbb66fa63572e).

`cf push` and `cf restart` will fail with this app. After a (few) restarts, the app will come up due to the Envoy becoming available.

@phil9909

**Comment by loewenstein** [comment]
@tcdowney Looking at the tracker stories, this will be fixable with Istio 1.7. Looking further into tracker, Istio 1.7 adoption is currently being worked on.

Can we expect to see Istio 1.7 and a fix for this in cf-for-k8s 0.7.0?
cc @keshav-pivotal @paulcwarren

**Comment by kauana** [comment]
Hey @loewenstein, @tcdowney is no longer on the networking team anymore. As you saw we are working on bumping istio to 1.7.1. We also have [another story](https://www.pivotaltracker.com/n/projects/2407973/stories/174125593) to check that the new istio version indeed fixes this issue.

Thanks

cc @XanderStrike @shalako
